modin/core/execution/ray/common/deferred_execution.py

```diff
@@ -0,0 +1,815 @@
+        self,
+        data: Union[
+            ObjectRefType,
+            "DeferredExecution",
+            List[Union[ObjectRefType, "DeferredExecution"]],
+        ],
+        func: Union[Callable, ObjectRefType],
+        args: Union[List[Any], Tuple[Any]],
+        kwargs: Dict[str, Any],
+        num_returns=1,
+    ):
+        ref = DeferredExecution._ref
+        ref(data)
+        self.data = data
+        self.func = func
+        self.args = args
+        self.kwargs = kwargs
+        self.num_returns = num_returns
+        self.flat_args = sum(ref(a) for a in args) == 0
+        self.flat_kwargs = sum(ref(a) for a in kwargs.values()) == 0
+        self.ref_counter = 1
+
+    @staticmethod
+    def _ref(obj):
+        """
+        Increment the `ref_counter` if `obj` is a `DeferredExecution`.
+
+        If `obj` is a `ListOrTuple`, this method is called recursively for each element.
+
+        Parameters
+        ----------
+        obj : Any
+
+        Returns
+        -------
+        int
+            The number of `ListOrTuple` or `DeferredExecution` objects found.
+        """
+        if isinstance(obj, DeferredExecution):
+            obj.ref_count(1)
+            return 1
+        if isinstance(obj, ListOrTuple):
+            ref = DeferredExecution._ref
+            result = 1
+            for o in obj:
+                result += ref(o)
+            return result
+        return 0
+
+    def exec(
+        self,
+    ) -> Tuple[ObjectRefOrListType, Union["MetaList", List], Union[int, List[int]]]:
+        """
+        Execute this task, if required.
+
+        Returns
+        -------
+        tuple
+            The execution result, MetaList, containing the length, width and
+            the worker's ip address (the last value in the list) and the values
+            offset in the list. I.e. length = meta_list[offset],
+            width = meta_list[offset + 1], ip = meta_list[-1].
+        """
+        if self.has_result:
+            return self.data, self.meta, self.meta_offset
+
+        if (
+            not isinstance(self.data, DeferredExecution)
+            and self.flat_args
+            and self.flat_kwargs
+            and self.num_returns == 1
+        ):
+            result, length, width, ip = remote_exec_func.remote(
+                self.func, self.data, *self.args, **self.kwargs
+            )
+            meta = MetaList([length, width, ip])
+            self._set_result(result, meta, 0)
+            return result, meta, 0
+
+        consumers, output = self._deconstruct()
+        # The last result is the MetaList, so adding +1 here.
+        num_returns = sum(c.num_returns for c in consumers) + 1
+        results = self._remote_exec_chain(num_returns, *output)
+        meta = MetaList(results.pop())
+        meta_offset = 0
+        results = iter(results)
+        for de in consumers:
+            if de.num_returns == 1:
+                de._set_result(next(results), meta, meta_offset)
```

Comment:
Why assign the entire list (`meta`) for each partition (partition ~ DeferredExecution), if we can only assign the parts directly related to it? It's possible to calculate the required part using offset.

Reply:
Because it's not materialized here.